src/main/scala/com/microsoft/hyperspace/index/sources/iceberg/IcebergRelation.scala

```diff
@@ -0,0 +1,286 @@
+        val basePathOpt = partitionBasePath.map(p => Map("basePath" -> p)).getOrElse(Map.empty)
+        val opts = plan.options - "path" +
+          ("snapshot-id" -> currentSnapshot.snapshotId().toString) +
+          ("as-of-timestamp" -> currentSnapshot.timestampMillis().toString) ++
+          basePathOpt
+
+        Relation(
+          Seq(
+            PathUtils
+              .makeAbsolute(table.location(), spark.sessionState.newHadoopConf())
+              .toString),
+          Hdfs(sourceDataProperties),
+          reader.readSchema().json,
+          fileFormatName,
+          opts)
+    }
+  }
+
+  /**
+   * Returns whether the current relation has parquet source files or not.
+   *
+   * @return Always true since Iceberg table files are stored as Parquet.
+   */
+  override def hasParquetAsSourceFormat: Boolean = true
+
+  /**
+   * Returns list of pairs of (file path, file id) to build lineage column.
+   *
+   * File paths should be the same format as "input_file_name()" of the given relation type.
+   * For [[IcebergRelation]], each file path should be in this format:
+   *   `file:/path/to/file`
+   *
+   * @param fileIdTracker [[FileIdTracker]] to create the list of (file path, file id).
+   * @return List of pairs of (file path, file id).
+   */
+  override def lineagePairs(fileIdTracker: FileIdTracker): Seq[(String, Long)] = {
+    fileIdTracker.getFileToIdMap.toSeq.map { kv =>
+      (kv._1._1, kv._2)
+    }
+  }
+
+  /**
+   * Options of the current relation.
+   */
+  override def options: Map[String, String] = plan.options
+
+  /**
+   * The partition schema of the current relation.
+   */
+  override def partitionSchema: StructType = plan.source match {
+    case _: IcebergSource =>
+      val dsOpts = new DataSourceOptions(plan.options.asJava)
+      val table = getIcebergTable
+      SparkSchemaUtil.convert(table.spec().schema())
+  }
+
+  /**
+   * Creates [[HadoopFsRelation]] based on the current relation.
+   *
+   * This is mainly used in conjunction with [[createLogicalRelation]].
+   */
+  override def createHadoopFsRelation(
+      location: FileIndex,
+      dataSchema: StructType,
+      options: Map[String, String]): HadoopFsRelation = plan.source match {
+    case _: IcebergSource =>
+      HadoopFsRelation(
+        location,
+        new StructType(),
+        dataSchema,
+        None,
+        new ParquetFileFormat,
+        options + IndexConstants.INDEX_RELATION_IDENTIFIER)(spark)
+  }
+
+  /**
+   * Creates [[LogicalRelation]] based on the current relation.
+   *
+   * This is mainly used to read the index files.
+   */
+  override def createLogicalRelation(
+      hadoopFsRelation: HadoopFsRelation,
+      newOutput: Seq[AttributeReference]): LogicalRelation = {
+    val updatedOutput =
+      newOutput.filter(attr => hadoopFsRelation.schema.fieldNames.contains(attr.name))
+    new LogicalRelation(hadoopFsRelation, updatedOutput, None, false)
+  }
+
+  private def getIcebergTable: Table = {
```

Comment:
nit: can this be just `private lazy val icebergTable: Table` so that we don't have to do `val t = getIcebergTable` on the caller side?

Reply:
There are some Iceberg options that may change from other writers and in this case it is desired to get the latest table instance.

Reply:
ah ok, make sense. Shall we rename this to `loadIcebergTable` to be explicit about it? "load" sounds more like it's getting the latest version than "get" does.